### DISCUSSION TITLE
A Huge Step Forward (goodbye ExtUI)

### DISCUSSION THREAD

**Discussion by Jyers** [Jyers/Marlin#807] [General Discussion]
My apologies for the clickbait title, but to be fair, it is accurate.
Anyone who pay's attention to the commit history may have been surprised to see two massive commits labeled "The big Move"
In order to avoid confusion and a bunch of questions I thought I'd clear it up briefly here.

The long and the short of it is that I have completely disconnected the screen code from Marlin's Extensible UI system. This may come as a shock to some, as at this point this firmware is know to most as "Ender 3 V2 Extensible UI" but I assure you I have valid reasoning for this move.

When I first moved to ExtUI, it seemed like the holy grail of Marlin lcd systems, giving me an easy framework to connect all my existing display functions to Marlin in a much better way. However, it would seem this was too good to be true. A week or so after moving to ExtUI, I was informed that it was not actually intended to be used for displays such as the V2 lcd. It is really meant for lcds with their own controllers, which only communicate with the mainboard through high level serial messages. The creality touch screens for instance utilize this system and are well suited to ExtUI. But by the time I learned this, I was too invested in the port to look for other options.

So this is where we've been until now. Utilizing an improper mechanism, but achieving great results. That is, with one caveat. An Extensible UI based dwin code would never be able to get merged upstream. This fact has been sitting in the back of my mind for quite some time, but I really didn't have the time or energy to make the switch. On top of this I wasn't even sure I'd be able to maintain the functionality I had attained under the scope of ExtUI. But I am happy to say, I managed it.

While I'm sure there are going to be some bugs and kinks in the first week, it's well worth knowing that this project is now finally on track to possibly being merged upstream. And if nothing else, this new framework should hopefully open the door to even more great features and improvements to come. 

TL;DR - The firmware no longer utilizes Marlin's Extensible UI framework, but that's not going to make it worse I promise. Also if you have any new catchy firmware name ideas let me know...

BIG CONFIG NOTE: `CREALITY_DWIN_EXTUI` is no longer being used. I have returned to the original `DWIN_CREALITY_LCD` as my new code has completely replaced my old rewrite. Custom icons are now `DWIN_CREALITY_LCD_CUSTOM_ICONS` Check the configuration templates for more info on the new defines. Also for the time being you need to change `ANY(HAS_MARLINUI_U8GLIB, HAS_MARLINUI_HD44780, IS_TFTGLCD_PANEL, EXTENSIBLE_UI)` and `EITHER(HAS_MARLINUI_U8GLIB, HAS_MARLINUI_HD44780, IS_TFTGLCD_PANEL, EXTENSIBLE_UI)` in configuration_adv.h to `ANY(HAS_MARLINUI_U8GLIB, HAS_MARLINUI_HD44780, IS_TFTGLCD_PANEL, EXTENSIBLE_UI, DWIN_CREALITY_LCD)` and `ANY(HAS_MARLINUI_U8GLIB, HAS_MARLINUI_HD44780, IS_TFTGLCD_PANEL, EXTENSIBLE_UI, DWIN_CREALITY_LCD)` These should just be temporary patches and I'll find a better way to address these soon.

**Comment by AngleV** [comment]
Finally, i was waiting for this.

As thinkyhead (the guy that maintain the upstream repo) stated somewhere that "Indeed, the ExtUI API is intended as a wrapper for displays that Marlin can't directly control, and which use a light protocol to send values to and from the controller. The Ender 3 V2 display is not suited to an ExtUI wrapper because it is entirely controlled by Marlin through custom code originally provided by Creality3D" i was sure that you would make the big leap.
That brings you to this well done desicion to go ahead and i am sure that you are going to succeed once more.
Let us know if we can provide any help somehow.

Wish you best luck and a good mood for porting this to MarlinUI implementation.

**Reply by joeclifford** [reply]
I was also waiting for this.

This gives a much greater chance of some of the main features (especial all the GUI improvements) making it is to the main Marlin branch.

I hoping the changes mriscoc has already had merged don't cause any issue with getting the Jyers improvments merged - https://github.com/MarlinFirmware/Marlin/pulls?q=is%3Apr+author%3Amriscoc

Mriscoc does seems to push his own varient heavily on facebook.

**Reply by Jyers** [reply]
If my code get's merged it will completely overwrite anything Mriscoc has done, as he kept working with the original code when I rewrote it. Which is unfortunate, but at least means his changes won't create any issues.

**Comment by VashBaldeus** [comment]
So all the features we've had with ExtUI are going to be gone or you are switching to a different method to maintain the features and progress further with development?

**Reply by Jyers** [reply]
I'm switching while maintaining the features and progressing with development.

**Comment by Tuncay-Ayhan** [comment]
@Jyers I'm pumped, can't wait to see what you are going to do with your FW. So no more restrictive choices? 

You can go all out I guess ? 😎 


People know your firmware as Jyers, so why shouldn't you change it do that? Slap the letters "UI" on it and you have JyersUI


**Ender_3_V2_Extensible_UI** evolves into .....


## _**JyersUI**_

**Comment by Reggi3** [comment]
@Jyers  Well done for making the switch, it can't have been an easy decision to make!  I can totally see why you would want to do this with your codebase, to anyone else wondering why, I hope this makes it easier to understand:
![image](https://user-images.githubusercontent.com/432746/118130847-ba5dc900-b3f5-11eb-905f-a7c0aa9745e8.png)

Keeping the code aligned with the original branch means we get all of the fixes that come online from all of the other contributors that are working directly with marlin, as well as Jyers top notch work for the E3v2.  

I am seeing that 4.2.x boards are turning up on a lot of machines, this move will hopefully make it much easier for people not using e3v2 screens to be able to have the same/similar experience?  I'm not entirely sure whether this is something specific that needs looking at by someone or whether it's easily doable right now?  I'm told that getting a vanilla Ender 3 screen working on this branch should be just some define changes in the configs but I'm not convinced, there seemed to be a lot of graphic stuff that is specific to our LCD?

**Comment by smooreace** [comment]
>
> this move will hopefully make it much easier for people not using e3v2
> screens to be able to have the same/similar experience
>

This is kinda what I was thinking. Ide love to move over to a much larger
touchscreen format! My eyes have taken a huge turn for the worse recently,
making it harder and harder to read the stock v2 display. The addition of
colors helped a TON though!

So here's to hoping that this new direction makes things a bit more
inclusive!

**Comment by rorywquin** [comment]
Can anyone explain this in non IT language.......please....

**Reply by Reggi3** [reply]
Sure :-)  Jyers marlin is a version of marlin developed from the original which covers 100+ printers, Jyers version is geared around 1 printer, the ender 3 v2.

During its development it has diverged from the original and the differences between the 2 make it difficult to merge the 2 versions back together so that they get the benefit of any new features/fixes that happen in either set of code.  Jyers is essentially fixing that so that his code aligns with the original marlin code, which means we get the best of both worlds, we get all of the features/fixes specifically for our printer AND we get any fixes and features that are developed by the original marlin crew too.

**Reply by rorywquin** [reply]
Cheers mate.

**Comment by smooreace** [comment]
There is a change on the horizon from coaxial to fiber networking

**Comment by Tapap84** [comment]
will any new releases still be viewable at https://github.com/Jyers/Marlin/releases

**Reply by Cal-LF** [reply]
@Tapap84 I am looking forward to the next update also. I have been compiling the updated 3X3 Manual Mesh template and there are many new features and bug fixes.

**Reply by gothcha** [reply]
Jacob, 1.3.3 compiled from source and working perfectly thank you. Once I found the changes required to my existing config files from your commits all was straightforward.
Love the blinking coords, even does each axis - neat! I will report any issues if they arise. Cheers :)

**Reply by Jyers** [reply]
@gothcha Thanks for the report! Happy to hear all is well :-)

**Reply by BejiitaImako** [reply]
Is this still gonna work as usual with the precompiled versions using only the bin file, no separate file for the screen needed?

**Reply by Jyers** [reply]
Yup!

**Comment by smooreace** [comment]
There never was a need to update your screen. The screen firmware is if you
want to modify the images on it. It has little to nothing to do with the
firmware directly

**Comment by BejiitaImako** [comment]
Great! Might be finally updating my machine next release then now that I
have made some other improvements to it getting it ready for more intense
and advanced production. My main focus for now is to make some sci fi style
LED lamps using transparent PETG in different colors to have as cool
looking mood/worklights for my balcony and a friends summer house / bbq
place nearby.



Den lör 5 juni 2021 21:37Jyers ***@***.***> skrev:
